### PR TITLE
create amenity JPA entity

### LINKED ISSUE #117: Add amenity JPA entity
## ⚠️ Is your feature request related to a problem? Please describe

* Add amenity JPA entity

## 💡 Describe the solution you'd like

* Add a new JPA entity `CommunityAmenity.java`

## Task Relationship

Subtask of: jmprathab/MyHome#93

### PR DESCRIPTION
**Pull request by DDmit04**
## 🚀 Description

Create amenity JPA entity

## 📄 Motivation and Context

Closes #117

### CODE REVIEW

File: src/main/java/com/myhome/domain/CommunityAmenity.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+public class CommunityAmenity extends BaseEntity {
+
+  @Column(nullable = false, unique = true)
+  private String amenityId;
+  @Column(nullable = false)
+  private String discription;
+  @Column
+  private boolean isBooked;
+  @Column
+  private LocalDate bookingStartDate;
```
Comment (jmprathab) [inline src/main/java/com/myhome/domain/CommunityAmenity.java]: This ideally has to be Date and Time.
Use case: HouseMember books a tennis court from 10 AM to 11 AM on 01-01-2020.

Comment (DDmit04) [reply src/main/java/com/myhome/domain/CommunityAmenity.java]: frontend can convert `01.01.2020 10 AM` to `01.01.2020 22:00:00`?

File: src/main/java/com/myhome/domain/CommunityAmenity.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  @Column
+  private boolean isBooked;
+  @Column
+  private LocalDateTime bookingStartDate;
+  @Column
+  private LocalDateTime bookingEndDate;
+  @ManyToOne
+  private Community community;
+
+  public boolean checkBooking() {
```
Comment (mslowiak) [inline src/main/java/com/myhome/domain/CommunityAmenity.java]: The entity should not have any logic

**Comment by mslowiak** [comment]
`Closing #117` - `Closes #117`

**Comment by DDmit04** [comment]
I also want to add `@EntityGraph` usage for hibernate relationship between `Community` and `CommnityAmenity`

**Comment by jmprathab** [comment]
> I also want to add `@EntityGraph` usage for hibernate relationship between `Community` and `CommnityAmenity`

Sure. Please go ahead.

**Comment by DDmit04** [comment]
`findByCommunityIdWithAmenities(String communityId)` not used anywhere but may be useful in the future

File: src/main/java/com/myhome/repositories/CommunityRepository.java [hunk 3]
```diff
@@ -27,8 +31,9 @@
 
   Optional<Community> findByCommunityId(String communityId);
 
-  @Transactional
-  Integer deleteByCommunityId(String communityId);
+  @Query("from Community comminity where comminity.communityId = :communityId")
+  @EntityGraph(value = "Community.amenities")
+  Optional<Community> findByCommunityIdWithAmenities(@Param("communityId") String communityId);
```
Comment (DDmit04) [inline src/main/java/com/myhome/repositories/CommunityRepository.java]: `findByCommunityIdWithAmenities` not used anywhere but may be useful in the future

**Comment by DDmit04** [comment]
And I head a question - why `mappedBy` in `@OneToMany` annotation is never used? If we use it, then we can make `data.sql` smaller

File: src/main/java/com/myhome/repositories/CommunityRepository.java (hunk 3 above)
Comment (mslowiak) [reply src/main/java/com/myhome/repositories/CommunityRepository.java]: If it is not used then it should be deleted :) 
https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it

**Comment by jmprathab** [comment]
> And I head a question - why `mappedBy` in `@OneToMany` annotation is never used? If we use it, then we can make `data.sql` smaller

True. We can do that. Till now, we have not prioritized it. But, this was one of the things in my mind.

File: src/main/java/com/myhome/repositories/CommunityRepository.java (hunk 3 above)
Comment (jmprathab) [reply src/main/java/com/myhome/repositories/CommunityRepository.java]: Typo on word `comminity`

**Comment by jmprathab** [comment]
@DDmit04 Let me review it in a day.

**Review by jmprathab — APPROVED** [review]
Looks good.

**Comment by jmprathab** [comment]
@mslowiak @DDmit04 @edwardUL99 

What do you think about removing the usage of `@AllArgsConstructor` from the codebase? I think it introduces the [COP problem](https://en.wikipedia.org/wiki/Connascence#Connascence_of_Position_(CoP)).

I personally am not a big fan of it and prefer hand-written all args constructor which gives us control over the order of the field and also to ignore specific fields from the constructor.

**Comment by edwardUL99** [comment]
> @mslowiak @DDmit04 @edwardUL99
> 
> What do you think about removing the usage of `@AllArgsConstructor` from the codebase? I think it introduces the [COP problem](https://en.wikipedia.org/wiki/Connascence#Connascence_of_Position_(CoP)).
> 
> I personally am not a big fan of it and prefer hand-written all args constructor which gives us control over the order of the field and also to ignore specific fields from the constructor.

@jmprathab I don't understand the issue with COP. If we wrote our own constructor, would we still not have to confirm to the order of the arguments? Or is the issue that you would prefer the choice of the order rather than the order being defined by the order they appear in the class definition?

I do agree with the fact we can ignore specific fields however

**Comment by jmprathab** [comment]
> @jmprathab I don't understand the issue with COP. If we wrote our own constructor, would we still not have to confirm to the order of the arguments? Or is the issue that you would prefer the choice of the order rather than the order being defined by the order they appear in the class definition?
> 
> I do agree with the fact we can ignore specific fields however

Lets say someone accidentally changed the order of members, it is very difficult to catch that at compile-time. More than that, it should be possible for us to omit certain variables(Mostly collections) so that whenever new variables get added some other place where the constructor is being called need not have to accommodate the variable by passing either `null` or `new HashSet<SomeObject>()` for instance.

**Comment by edwardUL99** [comment]
@jmprathab I don't see a problem removing them then :+1:

**Pull request merged**